File: server/pipeline/stage03StoryBible.ts
import { callStructured } from '../llm/client.ts';
import { Project, StoryBible } from '../types.ts';

export async function runStage03StoryBible(project: Project): Promise<Project> {
  const sot = project.source_of_truth;
  if (!sot) throw new Error('Source of Truth missing before Stage 03');

  const charSummary = sot.characters.map((c) => `${c.name} (${c.gender}, ${c.age}): ${c.appearance}`).join('\n');
  const eventSummary = sot.events.map((e) => `[Event ${e.order_index}] ${e.description} (${e.emotion})`).join('\n');
  const factsSummary = sot.facts.map((f) => `- ${f}`).join('\n');

  const prompt = `You are creating a comprehensive Story Bible for a cinematic adaptation.
The detected story language is: ${sot.detected_language}.
Please write the Story Bible in the detected story language (or authentic bilingual terms where appropriate).

Source Facts & Characters:
CHARACTERS:
${charSummary}

CHRONOLOGICAL EVENTS:
${eventSummary}

ATOMIC FACTS:
${factsSummary}

Generate the Story Bible JSON with:
{
  "title": "Compelling, evocative title for the video script",
  "summary": "2-3 polished sentences capturing the core premise, central conflict, and emotional resolution",
  "genre": "e.g. Mythological Fantasy, Historical Drama, Sci-Fi Thriller, Folk Tale, Romance, Action",
  "theme": "Core philosophical or moral message of the narrative",
  "important_facts": ["Key world rule 1", "Key world rule 2", "Key constraint 3"],
  "cultural_context": ["Cultural traditions, setting motifs, dialect notes, historical or mythical references"]
}`;

  const res = await callStructured<StoryBible>({
    prompt,
    stage: 3,
    stageName: 'Story Bible Generation',
    projectId: project.status.project_id,
    isLightTask: false,
  });

  const storyBible: StoryBible = {
    title: res.title || 'Cinematic Story Script',
    summary: res.summary || 'A compelling cinematic journey through timeless storytelling.',
    genre: res.genre || 'Cinematic Drama',
    theme: res.theme || 'Courage, Destiny, and Truth',
    important_facts: Array.isArray(res.important_facts) ? res.important_facts : sot.facts.slice(0, 5),
    cultural_context: Array.isArray(res.cultural_context) ? res.cultural_context : [],
    language: sot.detected_language,
  };

  project.story_bible = storyBible;
  project.stage_outputs['stage_03'] = storyBible;

  return project;
}

File: backend/pipeline/stage03_story_bible.py
from backend.llm.client import call_structured
from backend.models.schemas import Project, StoryBible

def run_stage03_story_bible(project: Project) -> Project:
    sot = project.source_of_truth
    if not sot:
        raise ValueError("Source of Truth missing before Stage 03")

    char_summary = "\n".join([f"{c.name} ({c.gender}, {c.age}): {c.appearance}" for c in sot.characters])
    event_summary = "\n".join([f"[Event {e.order_index}] {e.description} ({e.emotion})" for e in sot.events])
    facts_summary = "\n".join([f"- {f}" for f in sot.facts])

    prompt = f"""You are creating a comprehensive Story Bible for a cinematic adaptation.
The detected story language is: {sot.detected_language}.
Please write the Story Bible in the detected story language (or authentic bilingual terms where appropriate).

Source Facts & Characters:
CHARACTERS:
{char_summary}

CHRONOLOGICAL EVENTS:
{event_summary}

ATOMIC FACTS:
{facts_summary}

Generate the Story Bible JSON with:
{{
  "title": "Compelling, evocative title for the video script",
  "summary": "2-3 polished sentences capturing the core premise, central conflict, and emotional resolution",
  "genre": "e.g. Mythological Fantasy, Historical Drama, Sci-Fi Thriller, Folk Tale, Romance, Action",
  "theme": "Core philosophical or moral message of the narrative",
  "important_facts": ["Key world rule 1", "Key world rule 2", "Key constraint 3"],
  "cultural_context": ["Cultural traditions, setting motifs, dialect notes, historical or mythical references"]
}}"""

    res = call_structured(
        prompt=prompt,
        stage=3,
        stage_name="Story Bible Generation",
        project_id=project.status.project_id,
        is_light_task=False,
    )

    res_dict = res if isinstance(res, dict) else {}
    story_bible = StoryBible(
        title=res_dict.get("title") or "Cinematic Story Script",
        summary=res_dict.get("summary") or "A compelling cinematic journey through timeless storytelling.",
        genre=res_dict.get("genre") or "Cinematic Drama",
        theme=res_dict.get("theme") or "Courage, Destiny, and Truth",
        important_facts=res_dict.get("important_facts") if isinstance(res_dict.get("important_facts"), list) else sot.facts[:5],
        cultural_context=res_dict.get("cultural_context") if isinstance(res_dict.get("cultural_context"), list) else [],
        language=sot.detected_language,
    )

    project.story_bible = story_bible
    project.stage_outputs["stage_03"] = story_bible.model_dump()

    return project
